### PR TITLE
Avoid early compilation when initializing a RawModule instance

### PR DESCRIPTION
**Pull request by leofang**
Fix https://github.com/cupy/cupy/pull/3486#issuecomment-653145880. There, the CI error occurred because `RawModule` compiles at initialization, not when `get_function()` is called, so in a Read the Docs environment where nvcc is not available, the build fails. Now it's fixed by ~~checking if the build environment has CUDA or not~~ (**UPDATE**) deferring the compilation to when something needs to be fetched.

~~In fact, making this `no_cuda` flag (and potentially a few other ones) available could help some other Cython modules to conditionally compile.~~

cc: @cjnolet

### CODE REVIEW

**Comment by leofang** [comment]
Will do a few cosmetic changes.

**Comment by leofang** [comment]
Done.

**Comment by asi1024** [comment]
How about stopping compiling the kernel at `__init__` for the consistency with the behavior of `RawKernel`?

**Comment by leofang** [comment]
Good point. If no compatible is acceptable then sounds good. So after the change the compilation will happen when any method is called.

**Comment by leofang** [comment]
@asi1024 Done. PR title is also changed to reflect the update.

File: cupy_setup_build.py [hunk 1]
```diff
@@ -728,7 +731,7 @@ def cythonize(extensions, arg_options):
     # Embed signatures for Sphinx documentation.
     directives['embedsignature'] = True
 
-    cythonize_option_keys = ('annotate',)
+    cythonize_option_keys = ('annotate', 'compile_time_env')
```
Comment (asi1024) [inline cupy_setup_build.py:734]: This fix looks not needed anymore.

Comment (leofang) [reply cupy_setup_build.py:734]: OK I'll take it off from this PR and revive it in another (was planning to touch the nccl module using this capability).

**Comment by leofang** [comment]
> Recompile does not happen in this test now.

Good catch. I fixed it.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**